swap/swap.go

```diff
@@ -1,98 +1,451 @@
-// This file is part of the go-ethereum library.
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
 //
-// The go-ethereum library is free software: you can redistribute it and/or modify
+// The Swarm library is free software: you can redistribute it and/or modify
 // it under the terms of the GNU Lesser General Public License as published by
 // the Free Software Foundation, either version 3 of the License, or
 // (at your option) any later version.
 //
-// The go-ethereum library is distributed in the hope that it will be useful,
+// The Swarm library is distributed in the hope that it will be useful,
 // but WITHOUT ANY WARRANTY; without even the implied warranty of
 // MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 // GNU Lesser General Public License for more details.
 //
 // You should have received a copy of the GNU Lesser General Public License
-// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
 
 package swap
 
 import (
+	"context"
+	"crypto/ecdsa"
+	"encoding/json"
 	"errors"
 	"fmt"
+	"math/big"
 	"strconv"
 	"sync"
+	"time"
 
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethersphere/swarm/contracts/swap"
+	contract "github.com/ethersphere/swarm/contracts/swap"
 	"github.com/ethersphere/swarm/log"
 	"github.com/ethersphere/swarm/p2p/protocols"
 	"github.com/ethersphere/swarm/state"
 )
 
-// SwAP Swarm Accounting Protocol
+// ErrInvalidChequeSignature indicates the signature on the cheque was invalid
+var ErrInvalidChequeSignature = errors.New("invalid cheque signature")
+
+// Swap represents the SwAP Swarm Accounting Protocol
 // a peer to peer micropayment system
 // A node maintains an individual balance with every peer
 // Only messages which have a price will be accounted for
 type Swap struct {
-	stateStore state.Store        //stateStore is needed in order to keep balances across sessions
-	lock       sync.RWMutex       //lock the balances
-	balances   map[enode.ID]int64 //map of balances for each peer
+	api                 PublicAPI
+	store               state.Store          // store is needed in order to keep balances and cheques across sessions
+	lock                sync.RWMutex         // lock the store
+	balances            map[enode.ID]int64   // map of balances for each peer
+	cheques             map[enode.ID]*Cheque // map of cheques for each peer
+	peers               map[enode.ID]*Peer   // map of all swap Peers
+	backend             contract.Backend     // the backend (blockchain) used
+	owner               *Owner               // contract access
+	params              *Params              // economic and operational parameters
```

Comment:
also, pointer here might infer you'd like to change the `Params` on the fly. needed?

Reply:
Indeed currently we would not need a pointer. The current only member is initialized from a constant.

However, the idea of `Params` is to encapsulate economic parameters which are probable to be changed on the fly. For example, the amount at which to cash in, trigger thresholds to initiate topups of the account, even trigger tresholds for alarms etc.

The guidelines state "If in doubt, use a pointer". As this is most likely going to be extended, I think it is justified to keep it a pointer.